### PR TITLE
Make Tooltip upgradable

### PR DESCRIPTION
**Pull request by reaperrr**
This allows the tooltip to change via upgrades, for example when an actor is deployed or reaches elite status.

### CODE REVIEW

File: OpenRA.Game/Traits/Player/FrozenActorLayer.cs [hunk 1]
```diff
@@ -80,7 +80,7 @@ public FrozenActor(Actor self, PPos[] footprint, Shroud shroud, bool startsRevea
 			Bounds = self.Bounds;
 			TargetTypes = self.GetEnabledTargetTypes().ToHashSet();
 
-			tooltip = self.TraitsImplementing<ITooltip>().FirstOrDefault();
+			tooltip = self.TraitsImplementing<ITooltip>().FirstOrDefault(Exts.IsTraitEnabled);
```
Comment (pchote) [inline OpenRA.Game/Traits/Player/FrozenActorLayer.cs]: You'll need to store all of these here, and then pick the appropriate one in `RefreshState`.

File: mods/ts/rules/nod-vehicles.yaml [hunk 2]
```diff
@@ -73,6 +73,13 @@ TTNK:
 	Tooltip:
 		Name: Tick Tank
 		Description: Nod's main battle tank.\nCan deploy to gain extra protection.\n  Strong vs Vehicles\n  Weak vs Infantry, Aircraft
+		UpgradeTypes: deployed
+		UpgradeMaxEnabledLevel: 0
+	Tooltip@DEPLOYED:
+		Name: Deployed Tick Tank
```
Comment (pchote) [inline mods/ts/rules/nod-vehicles.yaml]: "Tick Tank (deployed)" please

Comment (MustaphaTR) [inline mods/ts/rules/nod-vehicles.yaml]: Deployed Tick Tank is what vanilla TS uses. But i think Tick Tank (deployed)  is better too.

**Comment by reaperrr** [comment]
Updated.

File: mods/ts/rules/nod-vehicles.yaml [hunk 3]
```diff
@@ -73,6 +73,13 @@ TTNK:
 	Tooltip:
 		Name: Tick Tank
 		Description: Nod's main battle tank.\nCan deploy to gain extra protection.\n  Strong vs Vehicles\n  Weak vs Infantry, Aircraft
+		UpgradeTypes: deployed
+		UpgradeMaxEnabledLevel: 0
+	Tooltip@DEPLOYED:
+		Name: Tick Tank (deployed)
+		Description: Nod's main battle tank.\nDeployed to gain extra protection.\n  Strong vs Vehicles\n  Weak vs Infantry, Aircraft
```
Comment (pchote) [inline mods/ts/rules/nod-vehicles.yaml]: This highlights the bogosity of having this on `Tooltip` instead of `Buildable`.  These alternate descriptions will never be used.  I won't ask you to do it here (it belongs in its own PR), but we really need to move this to the proper trait.

Comment (reaperrr) [inline mods/ts/rules/nod-vehicles.yaml]: Filed #12045 so we don't forget.

**Comment by reaperrr** [comment]
I think it's better to get #12141 merged first.

**Comment by reaperrr** [comment]
Rebased.

File: OpenRA.Mods.Common/Widgets/ObserverProductionIconsWidget.cs [hunk 4]
```diff
@@ -91,7 +91,7 @@ public override void Draw()
 
 				var rsi = actor.TraitInfo<RenderSpritesInfo>();
 				var icon = new Animation(world, rsi.GetImage(actor, world.Map.Rules.Sequences, faction));
-				icon.Play(actor.TraitInfo<TooltipInfo>().Icon);
+				icon.Play(actor.TraitInfos<TooltipInfo>().First(Exts.IsTraitEnabled).Icon);
```
Comment (reaperrr) [inline OpenRA.Mods.Common/Widgets/ObserverProductionIconsWidget.cs]: I was just wondering, shouldn't we move `Icon` to `Buildable` as well?

**Comment by reaperrr** [comment]
Filed #12258 which should be merged first, too.

File: OpenRA.Mods.Common/Traits/World/EditorActorPreview.cs [hunk 5]
```diff
@@ -70,7 +70,9 @@ public EditorActorPreview(WorldRenderer worldRenderer, string id, ActorReference
 				Footprint = new ReadOnlyDictionary<CPos, SubCell>(footprint);
 			}
 
-			var tooltip = Info.TraitInfoOrDefault<EditorOnlyTooltipInfo>() as TooltipInfoBase ?? Info.TraitInfoOrDefault<TooltipInfo>();
+			var tooltip = Info.TraitInfos<EditorOnlyTooltipInfo>().FirstOrDefault() as TooltipInfoBase
+				?? Info.TraitInfos<TooltipInfo>().FirstOrDefault();
```
Comment (reaperrr) [inline OpenRA.Mods.Common/Traits/World/EditorActorPreview.cs]: Note to self: This should probably be `.First(Exts.IsTraitEnabled)` too, in any case it should match `ActorSelectorLogic`.

**Comment by abcdefg30** [comment]
The dependencies are merged now.

**Comment by reaperrr** [comment]
Rebased and updated.

File: OpenRA.Game/Traits/Player/FrozenActorLayer.cs [hunk 6]
```diff
@@ -101,8 +101,9 @@ public void RefreshState()
 				DamageState = health.DamageState;
 			}
 
-			if (tooltip != null)
+			if (tooltips.Any(Exts.IsTraitEnabled))
```
Comment (pchote) [inline OpenRA.Game/Traits/Player/FrozenActorLayer.cs]: You can make this cleaner by pulling a `var tooltip = tooltips.FirstOrDefault(Exts.IsTraitEnabled)` outside the conditional and then keeping this as a null check.

File: OpenRA.Mods.Common/Traits/Tooltip.cs [hunk 7]
```diff
@@ -13,11 +13,11 @@
 
 namespace OpenRA.Mods.Common.Traits
 {
-	public abstract class TooltipInfoBase : ITraitInfo
+	public abstract class TooltipInfoBase : UpgradableTraitInfo
 	{
 		[Translate] public readonly string Name = "";
 
-		public abstract object Create(ActorInitializer init);
+		public override object Create(ActorInitializer init) { return this; }
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Tooltip.cs]: Returning `this` here is bogus (that is a traitinfo, not a trait).  You should be able to remove this line completely.

File: OpenRA.Mods.Common/Widgets/Logic/Ingame/ProductionTooltipLogic.cs [hunk 8]
```diff
@@ -115,7 +115,7 @@ static string ActorName(Ruleset rules, string a)
 		{
 			ActorInfo ai;
 			if (rules.Actors.TryGetValue(a.ToLowerInvariant(), out ai) && ai.HasTraitInfo<TooltipInfo>())
-				return ai.TraitInfo<TooltipInfo>().Name;
+				return ai.TraitInfos<TooltipInfo>().First(Exts.IsTraitEnabled).Name;
```
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/Logic/Ingame/ProductionTooltipLogic.cs]: It looks like this method is expected to return `a` if there isn't a valid tooltip, so better to use `FirstOrDefault` and only return if non-null here.

File: OpenRA.Mods.Common/Widgets/ViewportControllerWidget.cs [hunk 9]
```diff
@@ -191,7 +191,7 @@ public void UpdateMouseover()
 
 			if (underCursor != null)
 			{
-				ActorTooltip = underCursor.TraitsImplementing<ITooltip>().First();
+				ActorTooltip = underCursor.TraitsImplementing<ITooltip>().First(Exts.IsTraitEnabled);
```
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/ViewportControllerWidget.cs]: Crashing here if all tooltips are disabled isn't great (previously the check in `underCursor` would have prevented this.  It would be better to use `FirstOrDefault` and then only set `ActorTooltipExtra` / `TooltipType` if it is non-null.

**Review by pchote — CHANGES_REQUESTED** [review]

**Comment by reaperrr** [comment]
Updated accordingly.

In addition, I changed `Buildable` actors without `Tooltip` to not crash the game, but display the internal name on the build palette instead and throw a lint error.
I believe this is a bit more modder-friendly for mods in their early stages, as actors won't require `Tooltip` to be buildable.

**Comment by GraionDilach** [comment]
Please add an upgraded tooltip to the Sensor Array as well.

**Comment by reaperrr** [comment]
> Please add an upgraded tooltip to the Sensor Array as well.

Updated.

File: OpenRA.Mods.Common/Widgets/Logic/Ingame/ProductionTooltipLogic.cs [hunk 10]
```diff
@@ -115,7 +115,11 @@ static string ActorName(Ruleset rules, string a)
 		{
 			ActorInfo ai;
 			if (rules.Actors.TryGetValue(a.ToLowerInvariant(), out ai) && ai.HasTraitInfo<TooltipInfo>())
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Widgets/Logic/Ingame/ProductionTooltipLogic.cs]: Can't we drop the `HasTraitInfo` check now?

**Comment by abcdefg30** [comment]
Looks good to me otherwise. :+1: / :white_check_mark:

File: OpenRA.Mods.Common/Widgets/Logic/Ingame/ProductionTooltipLogic.cs (hunk 10 above)
Comment (reaperrr) [inline OpenRA.Mods.Common/Widgets/Logic/Ingame/ProductionTooltipLogic.cs]: Done.

**Comment by reaperrr** [comment]
Updated.

**Review by obrakmann — APPROVED** [review]
👍

**Comment by obrakmann** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/12729333016d5b6ef42830b9a8e135ce262fe830)
